### PR TITLE
Bluebird: Fix inference of filtered catch / Reduce redundancy

### PR DESCRIPTION
**Pull request by Retsam**
- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [X] Test the change in your own code. (Compile and run.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: http://bluebirdjs.com/docs/api/catch.html#filtered-catch
- [X] Increase the version number in the header if appropriate.

This relates to #17306: @well1791 @lhecker @jcready

### Background info:

Bluebird's `.catch` supports three different methods of filtering errors:

```javascript
somePromise()
    //By constructor
    .catch(SomeErrorType, e => {
         // e is an instance of SomeErrorType
    })
    //By predicate
    .catch(isSomeErrorType, e => {
        // e is an instance of SomeErrorType
    })
     //By object
    .catch({code: 400}, e => {
       // e is some object with code 400
    })
```

### The Problem

With the current typings, filtering by constructor doesn't infer correctly.  Instead of inferring `e` as an instance of `SomeError`, it's inferred as the `SomeError` constructor itself.  Instead, this currently needs to be written as:

```javascript
somePromise()
    .catch(SomeErrorType, (e: SomeErrorType) => { /* ... */ })
```

which isn't the worst thing ever, but it's certainly an annoyance.  (And, I suspect this is by far the most frequent use of the filters: it certainly is in my code)

The root of the problem seems to be that the constructor case is falling into the overloads for the object case.  Since the object case allows basically anything, the constructor is valid match.  I thought Typescript gave preference to overloads based on order, but this appears to not be the case, at least not in this instance.

### Proposed solution:

The current overloads are structured like:

```typescript
catch<E1 extends Error>(
  filter1: (new (...args: any[]) => E1), //Constructor case
  onReject: (error: E1) => R | PromiseLike<R>
): Bluebird<R>;
 
catch<E1>(
  filter1: ((error: E1) => boolean) | (E1 & object), //Predicate and object case
  onReject: (error: E1) => R | PromiseLike<R>
): Bluebird<R>;
```

(With equivalent versions for multiple filters, and for versions where `onReject` returns a type other than `R`)

I was able to get the inference to work as expected by consolidating these into:

```typescript
catch<E1>(
  filter1: (new (...args: any[]) => E1) | ((error: E1) => boolean) | (object & E1),
  onReject: (error: E1) => R | PromiseLike<R>
): Bluebird<R>;
```

The only downside to this change that I've found is that now *any* constructor is accepted as the first argument, rather than only accepting Error subtypes.  Personally, that's a trade-off that I'm willing to make, but certainly if there's a way to fix this without that compromise, that would be preferred, but after a fair bit of playing with this, I wasn't able to find it.  

---

Also, adds a commit that aliases `caught` as `catch`, without needing duplicate definitions.

### CODE REVIEW

**Comment by jcready** [comment]
> The only downside to this change that I've found is that now any constructor is accepted as the first argument, rather than only accepting Error subtypes.

This shouldn't be consider a downside IMO because you _can_ throw things other than instances of `Error`.

**Review by jcready — APPROVED** [review]

**Comment by Retsam** [comment]
> This shouldn't be consider a downside IMO because you can throw things other than instances of Error.

You can throw them, but you can't use their constructors to filter them.  For example this code will log "Normal catch":

```ts
class NotAnError {};

Promise.reject(new NotAnError())
    .catch(NotAnError, (e: NotAnError) => console.log("Filter catch"))
    .catch(() => console.log("Normal catch"))
```

Prior to this change the above code wouldn't compile, since `(e: NotAnError)` doesn't match the defined type of `(e: typeof NotAnError)` (again, it falls into the `object & E` case).  (But `.catch(NotAnError, e => console.log("Filtered catch"))` would have compiled, so it's not like the old behavior was bulletproof, either)

File: types/bluebird/bluebird-tests.ts [hunk 1]
```diff
@@ -352,7 +352,7 @@ class CustomError extends Error {
 	customField: number;
 }
 // $ExpectType Bluebird<void | Foo>
-fooProm.catch(CustomError, (reason: CustomError) => {
+fooProm.catch(CustomError, reason => {
```
Comment (Retsam) [inline types/bluebird/bluebird-tests.ts]: Changed this so that the tests will fail if this isn't inferred correctly.

**Comment by jcready** [comment]
Oh I see. I had no idea bluebird would limit the constructor in a filtered catch to only be `Error` instance producing ones. But I agree, this is still probably better than before.

File: types/bluebird/index.d.ts [hunk 2]
```diff
@@ -209,144 +152,7 @@ declare class Bluebird<R> implements PromiseLike<R>, Bluebird.Inspection<R> {
...
-  ): Bluebird<R>;
-  caught<U, E1 extends Error>(
-    filter1: (new (...args: any[]) => E1),
-    onReject: (error: E1) => U | PromiseLike<U>,
-  ): Bluebird<U | R>;
-  caught<U, E1>(
-    filter1: ((error: E1) => boolean) | (E1 & object),
-    onReject: (error: E1) => U | PromiseLike<U>,
-  ): Bluebird<U | R>;
+  caught: Bluebird<R>["catch"];
```
Comment (lhecker) [inline types/bluebird/index.d.ts]: Dude, this is f**** brilliant!
Of course… And it's such an obvious solution even… Why didn't we use that all the time yet?!

Mind Blown

**Review by lhecker — APPROVED** [review]
@Retsam Thanks again for this awesome PR! Please keep them coming. 😊

**Pull request merged**